src/workflows/zip/ZipWorkflowManager.ts

```diff
@@ -0,0 +1,182 @@
+  }
+
+  onTaskFinished(cb: (rs: Readable) => void): void {
+    this.ee.once('taskFinished', () => {
+      if (this.responseRS === null) {
+        this.ee.emit('error', 'Task finished without creating `responseRS`');
+
+        return;
+      }
+
+      cb(this.responseRS);
+    });
+  }
+
+  abort(): void {
+    this.ee.emit('abort');
+    if (this.responseRS !== null) {
+      this.responseRS.destroy();
+      this.responseRS = null;
+    }
+  }
+
+  private setupListeners() {
+    // If there were more than one errors keep logging them
+    this.ee.on('error', (err) => {
+      logger.logException(this.logPrefix + 'had error', err, {});
+    });
+
+    // if we encountered an error, we finished or were aborted
+    // try to stop any ongoing work and clean things up
+    this.ee.once('error', () =>
+      this.factories.forEach((factory) => factory.abort())
+    );
+    this.ee.once('abort', () =>
+      this.factories.forEach((factory) => factory.abort())
+    );
+    this.ee.once('taskFinished', () =>
+      this.factories.forEach((factory) => factory.cleanup())
+    );
+    this.ee.on('entityExecutionFinished', async () => {
+      if (this.factoriesMeta.size === this.data.length) {
+        await this.createZipFile()
+          .then((zipPath) => this.taskFinished(zipPath))
+          .catch((error) => this.ee.emit('error', error));
+      }
+    });
+  }
+  private createZipFile() {
+    const outPath = `${generateTmpPath()}.zip`;
+    const archive = archiver('zip', { zlib: { level: 9 } });
+    const stream = createWriteStream(outPath);
+
+    return new Promise<string>(async (resolve, reject) => {
+      stream.on('close', () => {
+        logger.logDebug(
+          `${
+            this.logPrefix
+          } zip file has been finalized and is ${archive.pointer()} total bytes`,
+          {}
+        );
+      });
+      archive.on('error', (err) => {
+        return reject(err);
+      });
+      for (let rootIdx = 0; rootIdx < this.data.length; rootIdx++) {
+        if (!this.factoriesMeta.has(rootIdx)) {
+          logger.logError(
+            this.logPrefix + `'${rootIdx}' is missing from 'factoriesMeta'`,
+            {
+              factoriesMeta: this.factoriesMeta,
+            }
+          );
+
+          throw new Error(`'${rootIdx}' is missing from 'factoriesMeta'`);
+        }
+
+        const { meta } = this.factoriesMeta.get(rootIdx)!;
+        for (const file of meta.files) {
+          archive.file(file.path, {
+            name: `/${meta.proposal}/${file.name}`,
+          });
+        }
+      }
+      archive.pipe(stream);
+      stream.on('close', () => {
+        return resolve(outPath);
+      });
+      archive.finalize();
+    });
+  }
+
+  private taskFinished(zipFilePath: string) {
+    this.responseRS = createReadStream(zipFilePath);
+    this.responseRS.on('error', (err) => this.ee.emit('error', err));
+    this.responseRS.once('close', () => failSafeDeleteFiles([zipFilePath]));
+
```

Comment:
I'm not sure how this error handling works, if zipFilePath is wrong path does is return null or?

Reply:
At this stage responseRS will start as null until createReadStream(zipFilePath) executes successfully.If zipFilePath is wrong path an error is thrown by createReadStream and we then emit that error so it can be handled.